Try drill down for pattern inserter

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Enables on large viewports the same drill down navigation we have in the site editor sidebar and in the small viewport for inserter categories, and in the process removes the preview pane.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The preview pane is not very useful: it creates another content area to skip for focus management, and it uses a lot of viewport area which is quite precious for assembling patterns (e.g. it should be possible on a large viewport to have both the inserter and the inspector open and assemble patterns). All that just to keep the category list available.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- Copies over the component used for the small viewport and adjusts it to also work on larger viewports.
- Removes some zoom out layout tweaks.
- Enables the assembling in zoom out mode when the pattern tab of the inserter is accessed (as opposed to when a category is accessed)
- Tries to make the media and patterns behave the same on category navigation

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Go to the site editor
2. Go to Pages and edit a page
3. Open the inserter
4. Open the patterns tab
5. **Notice zoom out mode is engaged**
6. **Notice the list of pattern categores**
7. Click a category
8. **Notice the patterns appear in the same panel**
9. Enable responsive mode
10. **Small viewport experience should be unchanged**


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

N/A

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/107534/ca660195-4876-4df4-8c12-180bab94ddbf

**Known Issues**

- [ ] There are some styling refinements required so that all panels look well spaced and similar
- [ ] There is a wild super scroll appearing when searching for patterns
- [ ] There is a sneaky disappearing scroll when the pattern list is rendering
- [ ] Somehow switching viewport size (?) may result in zoom out mode stuck to on
- [ ] We should move up the search in media if possible so that all tabs have the same UI structure

draganescu committed:
copy across viewports the mobile drill down navigation instead of dual panel for patterns

draganescu committed:
partial fix for the media panel on small viewports

draganescu committed:
use one column for search results

scruffian committed:
allow categories to go full width

draganescu committed:
unify the media and pattern behavior and implementation

draganescu committed:
simplify search result column

talldan commented:
Can probably adjust the padding around the patterns list, the left/right padding feels unnecessary and makes the previews much smaller than they need to be:
<img width="348" alt="Screenshot 2024-05-10 at 2 38 29 pm" src="https://github.com/WordPress/gutenberg/assets/677833/d81e6dc8-27fd-4d7a-a035-f406b16c3283">

Commit:
Minor style tweaks, add back class

annezazu commented:
Tested this out and the biggest pain point I found is getting back to all categories as once you start scrolling the back button disappears, along with the title of the category: 

https://github.com/WordPress/gutenberg/assets/26996883/7b0124cd-0078-46f6-bb1d-7314031d92fa

Since it's a fairly small section to return to, it's cumbersome to have to scroll all the way back up to see more.

jeryj committed:
Remove style hack as it's no longer needed

jeryj commented on packages/block-editor/src/components/inserter/block-patterns-tab/pattern-category-previews.js:
```diff
@@ -130,13 +132,25 @@ export function PatternCategoryPreviews( {
+							// TODO: This style override is also used in ToolsPanelHeader.
+							// It should be supported out-of-the-box by Button.
+							{ minWidth: 24, padding: 0 }
+						}
```
Looks like it is supported out of the box now. If we remove these lines the back button looks the same. Or maybe it was only in the context of the ToolsPanelHeader that it was an issue.

draganescu committed:
cleaning up the paddings of the list components, trying a new component name that makes more sense, moving content outside of header wrappers

jeryj committed:
Use arrow key navigation for inserterContentNavigator.

Todo:
- not sure about the roles. I don't think tree is right, but menu didn't seem right either
- Should probably pass an aria-label instead of fixing it as "categories"
- CSS cleanup

draganescu committed:
adds the experimental zoom out back

ellatrix commented:
I'd be good to get this in as soon as possible, so it can be used in #61489. Even on its own this is a nice change for 6.6.

Are there any blockers?

jameskoster commented:
I made the case for using the drilldown pattern when this was originally implemented, so it's a 👍 from me in general.

Another option could be something like to the block inserter, but with collapsible panels. Similar to Figma's asset browser.

<img width="397" alt="Screenshot 2024-05-21 at 15 28 23" src="https://github.com/WordPress/gutenberg/assets/846565/7eab943f-0908-40fb-828c-4965e1f23c08">


But on balance I suspect the drilldown would work better.

draganescu commented:
In terms of UI there is a discrepancy between how the search box works and how it is displayed between tab panels. 

Since we moved it inside each tab panel these two are to be solved:

1) should the search box always be the 1st item or should it be the second item after drill down header (back + title)? Should they differ between blocks, patterns and media to signify that for media it's not globally searching across categories?
2) what should be sticky on scroll? Both search and drill down header> Search too when a search is active?

draganescu committed:
add the media library button back

Commit:
Use small size for patterns filter + i18n

Commit:
Update metrics

jeryj committed:
Cleanup tech debt

The showPattern and showMedia and related constants were not needed in menu.js. Also, they were using children to render inner content which has also been removed.

jasmussen commented:
The main motivation is to free up space for the canvas to be visible when you're building with patterns, especially in context where the inserter stays open as you build. That's valid enough, and if those pieces come together, this drilldown mechanism can work for it. 

Here's trunk:

![before](https://github.com/WordPress/gutenberg/assets/1204802/db9ae106-e600-4031-b8de-26980d473408)

Here's this branch:

![after](https://github.com/WordPress/gutenberg/assets/1204802/8ab6051f-46b6-4722-b65e-7961ea25049d)

The issue Anne notes appears addressed. The main tradeoff here is that it's hard to quickly get a sense of all the patterns in the various categories. That's actually a biggish tradeoff, and it's not clear it's worth it to me. I do wonder, can we apply this drilldown pattern only to smaller viewport sizes?

If we do move forward, the media tab needs some love too:

![after](https://github.com/WordPress/gutenberg/assets/1204802/781bef4c-0d27-4b90-9132-e55f081ef604)

jasmussen commented:
Anecdotally, just today I went browsing to look for a nice pattern, and I had this branch checked out. I immediately switched to trunk because the clicking in and out was not a great way to browse.

ellatrix commented:
Imo, if we do this, the pattern category selector should be be a select box at the top vs a drill down. That would solve the problem @jasmussen is describing I think. We need to think data views more here?

jasmussen commented:
I'm not sure the dropdown would help. It's still two clicks and lacking the clear context of the category I'm already in. As noted, I don't know I'm suggesting this to be a blocker, it's ultimately a tradeoff, but at the moment I'm leaning towards finding ways to fit the inserter with the opened "tray" better on screen, than using the drilldown, or dropdown.

ellatrix commented:
@jameskoster The previews are probably a bit too small?

jameskoster commented:
Potentially yes, it will likely depend on the pattern/category. The number of columns could be easy to adjusted, even by the user (which is also something we want to add to data views: https://github.com/WordPress/gutenberg/issues/60630).

draganescu commented:
@jameskoster the previews in the data grid mockup you linked to are less than the two column previews from search results available in `trunk` which this PR _also removes_. 

For what we assume is an scanning behaviour to complete a set (to assemble the page with good patterns) having many small previews in a constrained space is less useful than having few large previews because it's very easy and common and expected to just scroll. 

I think the argument above is correct, but I do concede that having more than one in a view could be useful so it's not perfect. Yet, it's a good tradeoff considering the UX we aim for should [resemble assembling a stack](https://m.media-amazon.com/images/I/51V033EmnPL.jpg) rather than finding [waldo](https://i.sstatic.net/WEGrv.jpg) so each one larger is better than many of them smaller IMO.

jameskoster commented:
@draganescu I think it's a bit more nuanced; we should be aiming to create UIs that can adapt contextually, rather than optimising only for one specific flow. 

When you're initially building a page I agree the pattern browsing experience is better when the previews are arranged into a single column. But when you're searching for a specific pattern you've used before, a multi-column layout (or even a list) might be easier to scan. The benefit of using data views is that it empowers both the consumer and the user to craft a UI that best suits any given activity. 

Anyway, I appreciate it's a rabbit hole, and way beyond the scope of this PR :D

draganescu commented:
I will close this PR because as it stands it showed both the opportunity and the limitations of the UX. Given that the overall page pattern assembly is still being developed as an experiment we should probably tackle this as part of that work connected with those flows - unless the current UX proves tweak-able enough to become just right (*) 

* there was for instance an idea to partially scale the canvas so that the pane with previews for patterns and blocks does not overlap the content but instead by scaling the content makes room for the new panel.
